flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/PartitionCommitManager.java

```diff
@@ -0,0 +1,143 @@
+import static org.apache.flink.table.filesystem.FileSystemOptions.PARTITION_TIME_EXTRACTOR_CLASS;
+import static org.apache.flink.table.filesystem.FileSystemOptions.PARTITION_TIME_EXTRACTOR_TIMESTAMP_PATTERN;
+import static org.apache.flink.table.filesystem.FileSystemOptions.PARTITION_TIME_EXTRACTOR_KIND;
+import static org.apache.flink.table.filesystem.FileSystemOptions.PARTITION_TIME_INTERVAL;
+import static org.apache.flink.table.utils.PartitionPathUtils.extractPartitionValues;
+
+/**
+ * Manage partition and watermark information.
+ */
+public class PartitionCommitManager {
+
+	private static final ListStateDescriptor<Map<Long, Long>> WATERMARKS_STATE_DESC =
+			new ListStateDescriptor<>(
+					"checkpoint-id-to-watermark",
+					new MapSerializer<>(LongSerializer.INSTANCE, LongSerializer.INSTANCE));
+	private static final ListStateDescriptor<List<String>> PENDING_PARTITIONS_STATE_DESC =
+			new ListStateDescriptor<>(
+					"pending-partitions",
+					new ListSerializer<>(StringSerializer.INSTANCE));
+
+	private final ListState<Map<Long, Long>> watermarksState;
+	private final ListState<List<String>> pendingPartitionsState;
+	private final TreeMap<Long, Long> watermarks;
+	private final Set<String> pendingPartitions;
+	private final PartitionTimeExtractor extractor;
+	private final long timeIntervalMills;
+	private final List<String> partitionKeys;
+
+	public PartitionCommitManager(
+			boolean isRestored,
+			OperatorStateStore operatorStateStore,
+			ClassLoader userCodeClassLoader,
+			List<String> partitionKeys,
+			Configuration conf) throws Exception {
+		this.partitionKeys = partitionKeys;
+		String extractorKind = conf.get(PARTITION_TIME_EXTRACTOR_KIND);
+		String extractorClass = conf.get(PARTITION_TIME_EXTRACTOR_CLASS);
+		String extractorPattern = conf.get(PARTITION_TIME_EXTRACTOR_TIMESTAMP_PATTERN);
+		this.timeIntervalMills = conf.getOptional(PARTITION_TIME_INTERVAL)
+				.map(Duration::toMillis)
+				.orElse(Long.MAX_VALUE);
+		this.extractor = PartitionTimeExtractor.create(
+				userCodeClassLoader,
+				extractorKind,
+				extractorClass,
+				extractorPattern);
+
+		this.watermarksState = operatorStateStore.getListState(WATERMARKS_STATE_DESC);
+		this.pendingPartitionsState = operatorStateStore.getListState(PENDING_PARTITIONS_STATE_DESC);
+
+		this.watermarks = new TreeMap<>();
+		this.pendingPartitions = new HashSet<>();
+		if (isRestored) {
+			watermarks.putAll(watermarksState.get().iterator().next());
+			pendingPartitions.addAll(pendingPartitionsState.get().iterator().next());
+		}
+	}
+
+	/**
+	 * Add a pending partition.
+	 */
+	public void addPartition(String partition) {
+		if (!StringUtils.isNullOrWhitespaceOnly(partition)) {
+			this.pendingPartitions.add(partition);
+		}
+	}
+
+	/**
+	 * Trigger commit of pending partitions, and cleanup useless watermarks and partitions.
+	 */
+	public List<String> triggerCommit(long checkpointId) {
```

Comment:
I don't think this method really "triggers" the commit. Seems it just decides which partitions should be committed. So maybe rename to `getPartitionsToCommit`?

Reply:
`committablePartitions`